src/Compiler/Driver/GraphChecking/DependencyResolution.fs

```diff
@@ -156,25 +140,20 @@ let collectGhostDependencies (fileIndex: FileIndex) (trie: TrieNode) (queryTrie:
 
                 find trie path
 
-            let filesDefiningNamespace =
-                filesInTrie node |> Set.filter (fun idx -> idx < fileIndex)
-
-            let dependenciesDefiningNamespace =
-                Set.intersect result.FoundDependencies filesDefiningNamespace
-
-            [|
-                if Set.isEmpty dependenciesDefiningNamespace then
-                    // There is no existing dependency defining the namespace,
-                    // so we need to add one.
-                    if Set.isEmpty filesDefiningNamespace then
-                        // No file defines inferrable symbols for this namespace, but the namespace might exist.
-                        // Because we don't track what files define a namespace without any relevant content,
-                        // the only way to ensure the namespace is in scope is to add a link to every preceding file.
-                        yield! [| 0 .. (fileIndex - 1) |]
-                    else
-                        // At least one file defines the namespace - add a dependency to the first (top) one.
-                        yield Seq.head filesDefiningNamespace
-            |])
+            match node.Current with
+            // Both Root and module would expose data, so we can ignore them.
+            | Root _
+            | Module _ -> None
+            | Namespace (connectedFiles = connectedFiles) ->
+                if connectedFiles.Overlaps(result.FoundDependencies) then
+                    // The ghost dependency is already covered by a real dependency.
+                    None
+                else
+                    // We are only interested in any file that contained the namespace when they came before the current file.
+                    // If the namespace is defined in a file after the current file then there is no way the current file can reference it.
+                    // Which means that namespace would come from a different assembly.
+                    connectedFiles
+                    |> Seq.tryFind (fun connectedFileIdx -> connectedFileIdx < fileIndex))
```

Comment:
I think we want to retain the behaviour of linking to the top-most file that defines the namespace.
Is that guaranteed by the current code?
If so, it would be useful to make that intent more explicit, at least with a comment.